core/src/services/seafile/core.rs

```diff
@@ -0,0 +1,405 @@
+            }
+
+            let url = format!("{}/api2/repos", self.server);
+
+            let req = Request::get(url)
+                .header(header::AUTHORIZATION, format!("Token {}", auth_info.token))
+                .body(AsyncBody::Empty)
+                .map_err(new_request_build_error)?;
+
+            let resp = self.client.send(req).await?;
+
+            let status = resp.status();
+
+            match status {
+                StatusCode::OK => {
+                    let resp_body = &resp.into_body().bytes().await?;
+                    let list_library_response =
+                        serde_json::from_slice::<Vec<ListLibraryResponse>>(resp_body)
+                            .map_err(new_json_deserialize_error)?;
+
+                    for library in list_library_response {
+                        if library.name == self.repo_name {
+                            auth_info.repo_id = library.id;
+                            break;
+                        }
+                    }
+
+                    // repo not found
+                    if auth_info.repo_id.is_empty() {
+                        return Err(Error::new(
+                            ErrorKind::NotFound,
+                            &format!("repo {} not found", self.repo_name),
+                        ));
+                    }
+                }
+                _ => {
+                    return Err(parse_error(resp).await?);
+                }
+            }
+            Ok(auth_info.clone())
+        }
+    }
+}
+
+impl SeafileCore {
+    /// get upload url
+    pub async fn get_upload_url(&self) -> Result<String> {
+        let auth_info = self.get_auth_info().await?;
+
+        let req = Request::get(format!(
+            "{}/api2/repos/{}/upload-link/",
+            self.server, auth_info.repo_id
+        ));
+
+        let req = req
+            .header(header::AUTHORIZATION, format!("Token {}", auth_info.token))
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        let resp = self.send(req).await?;
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK => {
+                let resp_body = &resp.into_body().bytes().await?;
+                let upload_url = serde_json::from_slice::<String>(resp_body)
+                    .map_err(new_json_deserialize_error)?;
+                Ok(upload_url)
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    /// get download
+    pub async fn get_download_url(&self, path: &str) -> Result<String> {
+        let path = build_abs_path(&self.root, path);
+        let path = percent_encode_path(&path);
+
+        let auth_info = self.get_auth_info().await?;
+
+        let req = Request::get(format!(
+            "{}/api2/repos/{}/file/?p={}",
+            self.server, auth_info.repo_id, path
+        ));
+
+        let req = req
+            .header(header::AUTHORIZATION, format!("Token {}", auth_info.token))
```

Comment:
It's better to hide those signer related staff inside `SeafileSigner`.

Reply:
fix